src/nvim/buffer_defs.h

```diff
@@ -703,11 +703,10 @@ struct file_buffer {
                                 // may use a different synblock_T.
 
   struct {
-    int size;                   // last calculated number of sign columns
-    int max;                    // maximum value size is valid for.
-    linenr_T sentinel;          // a line number which is holding up the signcolumn
-    linenr_T invalid_top;       // first invalid line number that needs to be checked
-    linenr_T invalid_bot;       // last invalid line number that needs to be checked
+    int max;                    // maximum number of signs on a single line
+    int max_count;              // number of lines with max number of signs
+    bool resized;               // whether max changed at start of redraw
```

Comment:
Does this need to be state? It looks like this can be used as the return value for `buf_signcols_validate`, `buf_signcols_validate_range` and `buf_signcols_set_invalid`.

Reply:
Sure, it's only used to determine whether 'statuscolumn' needs to be rebuild when 'signcolumn' is "number/no". Because `w_scwidth` is and stays 0 then.

Reply:
Is it necessary to remove? It's more convenient as a globally accessible variable.

Reply:
Global state should generally be avoided, especially in a repository with a regular rotation of contributors who don't always 100% understand the code they have inherited. There are plenty of resources that explain this (quick one from google: https://softwareengineering.stackexchange.com/questions/148108/why-is-global-state-so-evil).

I know this isn't exactly global state, but given the size of the buf_T struct, the same principles all apply.

Sometimes the benefits of convenience can justify it, time is a real constraint after all, but I don't think that's the case here. I think removing this piece of state will make the code cleaner, even though it is a little more effort.

Reply:
Ok, so in this case it does actually need to be state.